### ISSUE TITLE
Compatibility issue between 5.5.0 and 5.6.0

### ISSUE THREAD

**Issue by sldab** [EsotericSoftware/kryo#1042]
I have a **RefreshTokenEntry** class which extends a class with the following attributes:

```
  String clientId;
  String nonce;
  URI redirectUri;
  List<String> scopeValues;
  boolean sessionless;
  boolean issuedForClient;
```

Serialization and deserialization worked with Kryo 5.5.0.
When trying to deserialize an object with Kryo 5.6.0 that was previously serialized with 5.5.0 I'm getting deserialization errors for the `redirectUri`.

**Stacktrace 1**
```
com.esotericsoftware.kryo.KryoException: java.net.URISyntaxException: Illegal character in path at index 42: tps://uat.xxxxxxx.dev/identity/signin-oidc^A^A  openid
Serialization trace:
redirectUri (de.xxx.rest.core.loginsession.RefreshTokenEntry)
        at com.esotericsoftware.kryo.serializers.DefaultSerializers$URISerializer.read(DefaultSerializers.java:1011)
        at com.esotericsoftware.kryo.serializers.DefaultSerializers$URISerializer.read(DefaultSerializers.java:1002)
        at com.esotericsoftware.kryo.Kryo.readObjectOrNull(Kryo.java:854)
        at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:133)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:130)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:880)
        ...
Caused by: java.net.URISyntaxException: Illegal character in path at index 42: tps://uat.xxxxxxx.dev/identity/signin-oidc^A^A   openid
        at java.base/java.net.URI$Parser.fail(URI.java:2976)
        at java.base/java.net.URI$Parser.checkChars(URI.java:3147)
        at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3229)
        at java.base/java.net.URI$Parser.parse(URI.java:3177)
        at java.base/java.net.URI.<init>(URI.java:623)
        at com.esotericsoftware.kryo.serializers.DefaultSerializers$URISerializer.read(DefaultSerializers.java:1009)
```

**Stacktrace 2**
```
com.esotericsoftware.kryo.KryoException: java.net.URISyntaxException: Illegal character in scheme name at index 0: .xxxxxxxx.yyyyyyyyyyyyy://login-callback/^A^A^Fopenid
Serialization trace:
redirectUri (de.xxx.rest.core.loginsession.RefreshTokenEntry)
        at com.esotericsoftware.kryo.serializers.DefaultSerializers$URISerializer.read(DefaultSerializers.java:1011)
        at com.esotericsoftware.kryo.serializers.DefaultSerializers$URISerializer.read(DefaultSerializers.java:1002)
        at com.esotericsoftware.kryo.Kryo.readObjectOrNull(Kryo.java:854)
        at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:133)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:130)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:880)
        ...
Caused by: java.net.URISyntaxException: Illegal character in scheme name at index 0: .xxxxxxxx.yyyyyyyyyyyyy://login-callback/^A^A^Fopenid
        at java.base/java.net.URI$Parser.fail(URI.java:2976)
        at java.base/java.net.URI$Parser.checkChars(URI.java:3147)
        at java.base/java.net.URI$Parser.checkChar(URI.java:3157)
        at java.base/java.net.URI$Parser.parse(URI.java:3172)
        at java.base/java.net.URI.<init>(URI.java:623)
        at com.esotericsoftware.kryo.serializers.DefaultSerializers$URISerializer.read(DefaultSerializers.java:1009)
        ... 38 common frames omitted
```

**Remarks**
* In both cases the deserializer seems to be off two bytes. 
* The scheme in stacktrace 1 is `https://`, not `tps://` and `openid` is already part of the `scopeValues` attribute, not part of the path.
* The custom scheme in stacktrace 2 is actually `de.xxxxxxxx.yyyyyyyyyyyyy://`, but it tries to parse it as `.xxxxxxxx.yyyyyyyyyyyyy://`.
* `nonce`, the attribute before `redirectUri`, is a random string.
* `issuedForClient` is a new attribute that didn't exist in the class when it was serialized.
* The migration worked without issues on another instance.

**JDK Version**
```
openjdk version "17.0.8.1" 2023-08-24
OpenJDK Runtime Environment Temurin-17.0.8.1+1 (build 17.0.8.1+1)
OpenJDK 64-Bit Server VM Temurin-17.0.8.1+1 (build 17.0.8.1+1, mixed mode, sharing)
```

**Comment by theigl** [comment]
Thanks for the detailed bug report @sldab!

There have been no changes to `URISerializer` between 5.5.0 and 5.6.0 and only 3 PRs that could potentially affect serialization: https://github.com/EsotericSoftware/kryo/releases/tag/kryo-parent-5.6.0

Do you use `setOptimizedGenerics(false)` in your Kryo config? If not, then the first PR does not affect you at all. The two other PRs only deal with `BigDecimalSerializer`.

Please try to write a reproducer test-case for this. Serialize `RefreshTokenEntry` with Kryo 5.5.0, copy the resulting bytes and try to deserialize the bytes into a new `RefreshTokenEntry` instance with the new field using Kryo 5.6.0.

**Comment by sldab** [comment]
Thanks for your quick response @theigl. 

I don't have `setOptimizedGenerics(false)` in my Kryo config.

I've downgraded to Kryo 5.5.0 and the issue still persisted, so it's not a compatiiblity issue as suspected.

Comparing a new record with an existing one revealed that the new record had two additional bytes before the affected fields, so the issue is likely caused by the new field. I've managed to manually patch the byte data of one of the records.

Yet, I don't understand why this issue didn't occur when upgrading the other instances of our system.
